lib/rails_param/param_evaluator.rb

```diff
@@ -0,0 +1,88 @@
+module RailsParam
+  class ParamEvaluator
+    attr_accessor :params
+
+    def initialize(params)
+      @params = params
+    end
+
+    def param!(name, type, options = {}, &block)
+      name = name.is_a?(Integer)? name : name.to_s
+      return unless params.include?(name) || check_param_presence?(options[:default]) || options[:required]
+
+      # coerce value
+      coerced_value = coerce(
+        params[name],
+        type,
+        options
+      )
+      parameter = RailsParam::Parameter.new(
+        name: name,
+        value: coerced_value,
+        type: type,
+        options: options,
+        &block
+      )
+
+      parameter.set_default if parameter.should_set_default?
+
+      # validate presence
+      if params[name].nil? && options[:required]
+        raise InvalidParameterError.new(
+          "Parameter #{name} is required",
+          param: name,
+          options: options
+        )
+      end
+
+      if block_given?
+        if parameter.type == Array
+          parameter.value.each_with_index do |element, i|
+            if element.is_a?(Hash) || element.is_a?(ActionController::Parameters)
+              recurse element, &block
+            else
+              parameter.value[i] = recurse({ i => element }, i, &block) # supply index as key unless value is hash
+            end
+          end
+        else
+          recurse parameter.value, &block
+        end
+      end
```

Comment:
I still feel like this bit should be delegated to the Array and Hash validators, but this can be done on a follow-up PR 👍